File: Britefury/DocModel/DMIOWrite.py
##-*************************
##-* This program is free software; you can use it, redistribute it and/or modify it
##-* under the terms of the GNU General Public License version 2 as published by the
##-* Free Software Foundation. The full text of the GNU General Public License
##-* version 2 can be found in the file named 'COPYING' that accompanies this
##-* program. This source code is (C)copyright Geoffrey French 1999-2007.
##-*************************
import pyparsing
import string

from Britefury.DocModel.DMListInterface import DMListInterface



"""
gSym document model IO


Uses basic S-expressions

lists are (...) as normal
tokens inside are:
	atom:   A-Z a-z 0.9 _+-*/%^&|!$@.,<>=[]~
	quoted string
	double quoted string
	another list
"""



## String IO helpers

def needsQuotes(s):
	return ' ' in s  or  '\t' in s  or  '\n' in s  or  '\r' in s  or  '\x0b' in s  or  '\x0c' in s  or  '(' in s  or  ')' in s  or  '`' in s  or  '{' in s  or  '}' in s  or  '\'' in s  or  len( s ) == 0






## WRITING FUNCTIONS

def _writeList(stream, content):
	stream.write( '(' )
	if len( content ) > 0:
		for v in content[:-1]:
			__writesx__( stream, v )
			stream.write( ' ' )
		__writesx__( stream, content[-1] )
	stream.write( ')' )


def __writesx__(stream, content):
	if isinstance( content, str )  or  isinstance( content, unicode ):
		if needsQuotes( content ):
			stream.write( repr( content ) )
		else:
			stream.write( content )
	elif isinstance( content, unicode ):
		stream.write( repr( content ) )
	elif isinstance( content, list ):
		_writeList( stream, content )
	elif isinstance( content, DMListInterface ):
		_writeList( stream, content )


def writeSX(stream, content):
	__writesx__( stream, content )






import unittest



class TestCase_DMIOWrite (unittest.TestCase):
	def _testWrite(self, data, result):
		import cStringIO
		stream = cStringIO.StringIO()
		writeSX( stream, data )
		value = stream.getvalue()
		if value != result:
			print 'ACTUAL RESULT:'
			print value
			print 'EXPECTED RESULT:'
			print result
		self.assert_( value == result )



	def testUnquotedString(self):
		self._testWrite( 'x', 'x' )
		
	def testQuotedString(self):
		self._testWrite( 'x y', '\'x y\'' )
		
	def testUnicodeString(self):
		from Britefury.DocModel.DMList import DMList
		self._testWrite( u'\u0107', 'u\'\\u0107\'' )
		
	def testEmptyString(self):
		self._testWrite( [ '' ], "('')" )
		

	def testWriteList(self):
		from Britefury.DocModel.DMList import DMList
		h = DMList()
		h.extend( [ 'h', '1', '2L', '3.0' ] )
		g = DMList()
		g.extend( [ 'g', h, 'Hi ' ] )
		f = DMList()
		f.extend( [ 'f', g, ' There', u'\u0107' ] )
		
		self._testWrite( f, '(f (g (h 1 2L 3.0) \'Hi \') \' There\' u\'\\u0107\')' )






if __name__ == '__main__':
	unittest.main()

File: java/BritefuryJ/DocModel/DMIOWrite.java
//##* This program is free software; you can use it, redistribute it and/or modify it
//##* under the terms of the GNU General Public License version 2 as published by the
//##* Free Software Foundation. The full text of the GNU General Public License
//##* version 2 can be found in the file named 'COPYING' that accompanies this
//##* program. This source code is (C)copyright Geoffrey French 2008.
//##************************
package BritefuryJ.DocModel;

import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import org.python.core.PyString;
import org.python.core.PyUnicode;



/*
 * gSym document model IO
 *
 *
 *Uses basic S-expressions

 *lists are (...) as normal
 *tokens inside are:
 *	atom:   A-Z a-z 0.9 _+-* /%^&|!$@.,<>=[]~
 *	quoted string
 *	double quoted string
 *	another list
 *
 */

public class DMIOWrite
{
	public static class InvalidDataTypeException extends Exception
	{
		private static final long serialVersionUID = 1L;
	}
	
	
	public static String unquotedStringPunctuationChars = "+-*/%^&|!$@.,<>=[]~";
	public static String quotedStringPunctuationChars = "+-*/%^&|!$@.,<>=[]~'()` ";
	public static Pattern unquotedString = Pattern.compile( "[0-9A-Za-z_" + Pattern.quote( unquotedStringPunctuationChars ) + "]+" );
	public static Pattern quotedStringContents = Pattern.compile( "[0-9A-Za-z_" + Pattern.quote( quotedStringPunctuationChars ) + "]+" );

	
	public static void escape(StringBuilder builder, String x)
	{
		for (int i = 0; i < x.length(); i++)
		{
			char c = x.charAt( i );
			
			if ( c == '\\' )
			{
				builder.append( "\\\\" );
			}
			else if ( c == '\n' )
			{
				builder.append( "\\n" );
			}
			else if ( c == '\r' )
			{
				builder.append( "\\r" );
			}
			else if ( c == '\t' )
			{
				builder.append( "\\t" );
			}
			else
			{
				builder.append(  "\\x" + Integer.toString( (int)c, 16 ) + "x" );
			}
		}
	}
	
	public static String quoteString(String s)
	{
		StringBuilder builder = new StringBuilder();
		
		builder.append( "\"" );
		// Escape newlines, CRs, tabs, and backslashes
		int pos = 0;
		
		while ( pos < s.length() )
		{
			Matcher m = quotedStringContents.matcher( s.substring( pos ) );
			
			if ( m.find()  &&  m.end() > m.start() )
			{
				if ( m.start() > 0 )
				{
					escape( builder, s.substring( pos, pos+m.start() ) );
				}
				String x = m.group();
				builder.append( x );
				pos += m.end();
			}
			else
			{
				escape( builder, s.substring( pos ) );
				break;
			}
		}
		
		builder.append( "\"" );

		return builder.toString();
	}

	
	public static void writeString(StringBuilder builder, String content)
	{
		Matcher m = unquotedString.matcher( content );
		boolean bFound = m.find();
		if ( bFound  &&  m.start() == 0  &&  m.end() == content.length() )
		{
			builder.append( content );
		}
		else
		{
			builder.append( quoteString( content ) );
		}
	}
	
	public static void writeList(StringBuilder builder, List<Object> content) throws InvalidDataTypeException
	{
		builder.append( "[" );
		if ( content.size() > 0 )
		{
			for (Object v: content.subList( 0, content.size() - 1 ))
			{
				writeSX( builder, v );
				builder.append( " " );
			}
			writeSX( builder, content.get( content.size() - 1 ) );
		}
		builder.append( "]" );
	}
	
	
	@SuppressWarnings("unchecked")
	public static void writeSX(StringBuilder builder, Object content) throws InvalidDataTypeException
	{
		if ( content instanceof String )
		{
			writeString( builder, (String)content );
		}
		else if ( content instanceof PyString )
		{
			writeString( builder, ((PyString)content).toString() );
		}
		else if ( content instanceof PyUnicode )
		{
			writeString( builder, ((PyUnicode)content).toString() );
		}
		else if ( content instanceof List )
		{
			writeList( builder, (List<Object>)content );
		}
		else
		{
			System.out.println( "Content data type: " + content.getClass().getName() + ", content data: " + content.toString() );
			throw new InvalidDataTypeException();
		}
	}



	public static String writeSX(Object content) throws InvalidDataTypeException
	{
		StringBuilder builder = new StringBuilder();
		writeSX( builder, content );
		return builder.toString();
	}
}


